### ISSUE TITLE
Is this a complete implementation of CNN?

### ISSUE THREAD

**Issue by xiaoiker** [TDeVries/enas_pytorch#1]
RT, is this a complete implementation? Do you get the reported performance on CIFAR and Imagenet?

**Comment by TDeVries** [comment]
I only implemented the macro architecture search for CNN (no micro search), but I tried to copy the official TensorFlow implementation as best as I could. I ran some tests but didn't spend too much time tuning the parameters, since I just used whatever values they used. It's been a while since I ran the tests (CIFAR-10 only), but if I recall correctly I got perhaps a percent worse than reported in the paper? It always seems to converge to using a single type of convolutional block, while in the paper the example architecture that they showed uses a combination of blocks, so I'm not sure why there is a difference. I think perhaps better performance could be achieved by doing a better job of fine-tuning the hyperparamters.

**Comment by MiZhangWhuer** [comment]
Hi, @xiaoiker @TDeVries , I have successfully run your code and I found the output of best_dag like this:
Best architecture:
[3]
[3 1]
[3 0 0]
[3 0 0 0]
[3 0 0 0 0]
[2 0 1 0 0 0]
[3 0 0 1 0 0 1]
[3 0 1 0 0 0 0 1]
[3 0 1 0 0 1 0 0 1]
[3 0 0 1 0 0 0 0 0 1]
[3 1 0 0 1 0 0 1 0 0 0]
[3 1 0 1 1 1 0 0 1 0 1 0]
What is the exact meaning of this matrix? What's the input and prior node presented in each line?
Do you mean that each line is the node id and each non-zero number corresponds to the index of prior nodes? 
If so, what's the prior nodes for [3] #line 1  and [3 0 0] #line 3 [3 0 0 0] #line 4? You mean that these nodes do not have prior input?

**Comment by TDeVries** [comment]
Yes, the first number in each row are the operation types, which are defined here: https://github.com/TDeVries/enas_pytorch/blob/a3e54b290787572a0fab920d9f3b3149e09dbc82/models/shared_cnn.py#L73

The following numbers indicate whether there is a skip connection between the current layer and the nth layer or not. Nodes in layer n-1 always connect to layer n, but there could also be an additional skip connection.

**Comment by TDeVries** [comment]
> RT, is this a complete implementation? Do you get the reported performance on CIFAR and Imagenet?

I'm actually not sure if I implemented the cosine learning rate properly, so if you aren't seeing the expected performance that might be something to look into.

**Comment by MiZhangWhuer** [comment]
> Yes, the first number in each row are the operation types, which are defined here:
> 
> [enas_pytorch/models/shared_cnn.py](https://github.com/TDeVries/enas_pytorch/blob/a3e54b290787572a0fab920d9f3b3149e09dbc82/models/shared_cnn.py#L73)
> 
> Line 73 in [a3e54b2](/TDeVries/enas_pytorch/commit/a3e54b290787572a0fab920d9f3b3149e09dbc82)
> 
>  self.branch_0 = ConvBranch(in_planes, out_planes, kernel_size=3) 
> The following numbers indicate whether there is a skip connection between the current layer and the nth layer or not. Nodes in layer n-1 always connect to layer n, but there could also be an additional skip connection.

If I understand correctly, you mean that for these two nodes (index start from 0) in the table :
[3 0 0] #line 3 (node 2) 
[3 0 0 0] #line 4 (node 3)
node 2 is the prior node of node 3 although node 3 does not have skip connection to other nodes?

Here is the code to show my opinion:

    for key, value in sample_arc.items():  
        if len(value) == 1:  
            branch_type = value[0].cpu().numpy()  
            add_cnnnode(graph, key, cnnop_list[branch_type[0]])  
        else:
            branch_type = value[0].cpu().numpy()  
            add_cnnnode(graph, key, cnnop_list[branch_type[0]]) 
            skips = value[1].cpu().numpy()  
            for i, skip in enumerate(skips):
                if skip == 1:
                   graph.add_edge(i, key)
                else:
                    graph.add_edge(key-1, key) 
Am I right?

**Comment by TDeVries** [comment]
> If I understand correctly, you mean that for these two nodes (index start from 0) in the table :
> [3 0 0] #line 3 (node 2)
> [3 0 0 0] #line 4 (node 3)
> node 2 is the prior node of node 3 although node 3 does not have skip connection to other nodes?

Yes, node 2 is the prior node of node 3 even though it has no skip connections. If instead we had
[3 0 0] #line 3 (node 2)
[3 0 0 1] #line 4 (node 3)
then node 2 would feed into node 3 _and_ there would be a skip connection.

>             if skip == 1:
>                graph.add_edge(i, key)
>             else:
>                 graph.add_edge(key-1, key) 
For this part I am not sure I follow. When skip == 1 it looks like it makes a skip connection between the ith layer and the current one, but what is it doing when skip == 0?

**Comment by MiZhangWhuer** [comment]
> > If I understand correctly, you mean that for these two nodes (index start from 0) in the table :
> > [3 0 0] #line 3 (node 2)
> > [3 0 0 0] #line 4 (node 3)
> > node 2 is the prior node of node 3 although node 3 does not have skip connection to other nodes?
> 
> Yes, node 2 is the prior node of node 3 even though it has no skip connections. If instead we had
> [3 0 0] #line 3 (node 2)
> [3 0 0 1] #line 4 (node 3)
> then node 2 would feed into node 3 _and_ there would be a skip connection.
> 
> > ```
> >         if skip == 1:
> >            graph.add_edge(i, key)
> >         else:
> >             graph.add_edge(key-1, key) 
> > ```
> 
> For this part I am not sure I follow. When skip == 1 it looks like it makes a skip connection between the ith layer and the current one, but what is it doing when skip == 0?

If skip == 0, it will connect to prior node index by key-1:

           else:
            graph.add_edge(key-1, key)
Here is my whole code to draw the best_dag:

    add_cnnnode(graph, -1, "image")
    maxkey = 0
    for key, value in sample_arc.items():
        graph.add_edge(str(int(key) - 1), key)
        if len(value) == 1:
            branch_type = value[0].cpu().numpy()
            print('branch_type', type(branch_type[0]))
            add_cnnnode(graph, key, cnnop_list[branch_type[0]])
        else:
            branch_type = value[0].cpu().numpy()
            add_cnnnode(graph, key, cnnop_list[branch_type[0]])
            skips = value[1].cpu().numpy()
            for i, skip in enumerate(skips):
                if skip == 1:
                   graph.add_edge(i, key)
        maxkey = key

    add_cnnnode(graph, str(int(maxkey)+1), "avg pool")
    graph.add_edge(str(int(maxkey)), str(int(maxkey)+1))
    add_cnnnode(graph, str(int(maxkey)+2), "dropout")
    graph.add_edge(str(int(maxkey)+1), str(int(maxkey)+2))
    add_node(graph, str(int(maxkey)+3), "fc")
    graph.add_edge(str(int(maxkey)+2), str(int(maxkey)+3))

    graph.layout(prog='dot')
    graph.draw(path)

By the way, are you still working on micro search algorithm? Hope your update on micro search & multi-gpu training on this excellent work. I've benefit a lot from your implementation.  It's really efficient compare to other implementation on github.

**Comment by TDeVries** [comment]
Okay, I think it looks correct. 

Sorry to say I am no longer working on this project.

**Comment by SuperRui41** [comment]
> > > If I understand correctly, you mean that for these two nodes (index start from 0) in the table :
> > > [3 0 0] #line 3 (node 2)
> > > [3 0 0 0] #line 4 (node 3)
> > > node 2 is the prior node of node 3 although node 3 does not have skip connection to other nodes?
> > 
> > 
> > Yes, node 2 is the prior node of node 3 even though it has no skip connections. If instead we had
> > [3 0 0] #line 3 (node 2)
> > [3 0 0 1] #line 4 (node 3)
> > then node 2 would feed into node 3 _and_ there would be a skip connection.
> > > ```
> > >         if skip == 1:
> > >            graph.add_edge(i, key)
> > >         else:
> > >             graph.add_edge(key-1, key) 
> > > ```
> > 
> > 
> > For this part I am not sure I follow. When skip == 1 it looks like it makes a skip connection between the ith layer and the current one, but what is it doing when skip == 0?
> 
> If skip == 0, it will connect to prior node index by key-1:
> 
> ```
>        else:
>         graph.add_edge(key-1, key)
> ```
> 
> Here is my whole code to draw the best_dag:
> 
> ```
> add_cnnnode(graph, -1, "image")
> maxkey = 0
> for key, value in sample_arc.items():
>     graph.add_edge(str(int(key) - 1), key)
>     if len(value) == 1:
>         branch_type = value[0].cpu().numpy()
>         print('branch_type', type(branch_type[0]))
>         add_cnnnode(graph, key, cnnop_list[branch_type[0]])
>     else:
>         branch_type = value[0].cpu().numpy()
>         add_cnnnode(graph, key, cnnop_list[branch_type[0]])
>         skips = value[1].cpu().numpy()
>         for i, skip in enumerate(skips):
>             if skip == 1:
>                graph.add_edge(i, key)
>     maxkey = key
> 
> add_cnnnode(graph, str(int(maxkey)+1), "avg pool")
> graph.add_edge(str(int(maxkey)), str(int(maxkey)+1))
> add_cnnnode(graph, str(int(maxkey)+2), "dropout")
> graph.add_edge(str(int(maxkey)+1), str(int(maxkey)+2))
> add_node(graph, str(int(maxkey)+3), "fc")
> graph.add_edge(str(int(maxkey)+2), str(int(maxkey)+3))
> 
> graph.layout(prog='dot')
> graph.draw(path)
> ```
> 
> By the way, are you still working on micro search algorithm? Hope your update on micro search & multi-gpu training on this excellent work. I've benefit a lot from your implementation. It's really efficient compare to other implementation on github.

Your discussion has benefited me a lot. May I know how to use your code to draw DAG? Where should it be added? How to realize it?

**Comment by SuperRui41** [comment]
> Okay, I think it looks correct.
> 
> Sorry to say I am no longer working on this project.

It's a pity.Thank you for your excellent implementation. I ran your code and got some results. This is shown below. What puzzles me is whether the code runs completely and correctly. As you can see, only “epoch 0” results are recorded in the “ENAS. Logs“ file.

```
Namespace(batch_size=128, child_grad_bound=5.0, child_keep_prob=0.9, child_l2_reg=0.00025, child_lr_T=10, child_lr_max=0.05, child_lr_min=0.0005, child_num_branches=6, child_num_layers=12, child_out_filters=36, controller_bl_dec=0.99, controller_entropy_weight=0.0001, controller_lr=0.001, controller_lstm_num_layers=1, controller_lstm_size=64, controller_num_aggregate=20, controller_op_tanh_reduce=2.5, controller_skip_target=0.4, controller_skip_weight=0.8, controller_tanh_constant=1.5, controller_train_every=1, controller_train_steps=50, cutout=0, data_path='/data/marui/ENAS/ENAS_macro_architecture/enas_pytorch-master//data/', eval_every_epochs=1, fixed_arc=False, log_every=50, num_epochs=310, output_filename='ENAS', resume='', search_for='macro', seed=0)
Files already downloaded and verified
Files already downloaded and verified
Files already downloaded and verified
epoch=0	ch_step=0	loss=2.335411	lr=0.0500	|g|=0.7008	acc=0.0938	time=2.05it/s
epoch=0	ch_step=50	loss=2.287261	lr=0.0500	|g|=0.7261	acc=0.1250	time=3.50it/s
epoch=0	ch_step=100	loss=2.197320	lr=0.0500	|g|=1.0307	acc=0.1797	time=5.73it/s
epoch=0	ch_step=150	loss=2.086418	lr=0.0500	|g|=0.6955	acc=0.2109	time=3.25it/s
epoch=0	ch_step=200	loss=2.144334	lr=0.0500	|g|=1.0483	acc=0.1016	time=5.81it/s
epoch=0	ch_step=250	loss=1.967500	lr=0.0500	|g|=0.5173	acc=0.2188	time=4.01it/s
epoch=0	ch_step=300	loss=2.083941	lr=0.0500	|g|=0.7872	acc=0.2344	time=3.26it/s
epoch=0	ch_step=350	loss=2.043149	lr=0.0500	|g|=0.5454	acc=0.1797	time=2.89it/s
Epoch 0: Training controller
ctrl_step=1	loss=-4.506	ent=67.22	lr=0.0010	|g|=0.3361	acc=0.2020	bl=0.27	time=2.37it/s
ctrl_step=3	loss=1.589	ent=67.04	lr=0.0010	|g|=0.2526	acc=0.2880	bl=0.27	time=2.70it/s
ctrl_step=5	loss=-0.158	ent=66.72	lr=0.0010	|g|=0.1435	acc=0.2645	bl=0.27	time=1.53it/s
ctrl_step=7	loss=-0.494	ent=66.24	lr=0.0010	|g|=0.1366	acc=0.2566	bl=0.26	time=2.26it/s
ctrl_step=9	loss=1.180	ent=65.94	lr=0.0010	|g|=0.3034	acc=0.2800	bl=0.26	time=1.98it/s
ctrl_step=11	loss=-0.890	ent=64.62	lr=0.0010	|g|=0.1471	acc=0.2486	bl=0.26	time=1.68it/s
ctrl_step=13	loss=3.134	ent=63.77	lr=0.0010	|g|=0.1585	acc=0.3033	bl=0.26	time=1.52it/s
ctrl_step=15	loss=0.194	ent=63.00	lr=0.0010	|g|=0.0463	acc=0.2563	bl=0.25	time=1.77it/s
ctrl_step=17	loss=0.339	ent=62.52	lr=0.0010	|g|=0.0625	acc=0.2563	bl=0.25	time=1.64it/s
ctrl_step=19	loss=1.554	ent=61.87	lr=0.0010	|g|=0.0840	acc=0.2718	bl=0.25	time=1.86it/s
ctrl_step=21	loss=-2.845	ent=61.24	lr=0.0010	|g|=0.0370	acc=0.1936	bl=0.24	time=1.69it/s
ctrl_step=23	loss=2.295	ent=61.40	lr=0.0010	|g|=0.1363	acc=0.2796	bl=0.24	time=1.77it/s
ctrl_step=25	loss=-1.666	ent=60.65	lr=0.0010	|g|=0.4706	acc=0.2092	bl=0.24	time=2.32it/s
ctrl_step=27	loss=-4.855	ent=62.75	lr=0.0010	|g|=0.1806	acc=0.1547	bl=0.24	time=2.12it/s
ctrl_step=29	loss=2.574	ent=61.64	lr=0.0010	|g|=0.0731	acc=0.2796	bl=0.24	time=1.84it/s
ctrl_step=31	loss=-4.894	ent=61.11	lr=0.0010	|g|=0.0617	acc=0.1545	bl=0.24	time=1.94it/s
ctrl_step=33	loss=-5.790	ent=60.59	lr=0.0010	|g|=0.0470	acc=0.1389	bl=0.24	time=1.90it/s
ctrl_step=35	loss=4.070	ent=60.95	lr=0.0010	|g|=0.1194	acc=0.2952	bl=0.23	time=2.72it/s
ctrl_step=37	loss=-0.573	ent=60.31	lr=0.0010	|g|=0.0381	acc=0.2170	bl=0.23	time=2.49it/s
ctrl_step=39	loss=2.920	ent=61.01	lr=0.0010	|g|=0.1801	acc=0.2795	bl=0.23	time=2.03it/s
ctrl_step=41	loss=3.127	ent=60.08	lr=0.0010	|g|=0.1822	acc=0.2794	bl=0.23	time=2.49it/s
ctrl_step=43	loss=1.278	ent=60.04	lr=0.0010	|g|=0.0655	acc=0.2482	bl=0.22	time=2.42it/s
ctrl_step=45	loss=2.988	ent=60.25	lr=0.0010	|g|=0.0295	acc=0.2795	bl=0.23	time=2.68it/s
ctrl_step=47	loss=2.333	ent=59.77	lr=0.0010	|g|=0.0440	acc=0.2716	bl=0.23	time=2.42it/s
ctrl_step=49	loss=-0.994	ent=59.73	lr=0.0010	|g|=0.0751	acc=0.2091	bl=0.23	time=2.05it/s
Here are 10 architectures:
[2]
[4 0]
[3 0 0]
[5 1 0 0]
[4 0 0 0 0]
[2 0 0 0 1 1]
[1 1 0 0 1 0 1]
[3 1 0 1 0 0 0 0]
[0 1 0 0 0 0 1 1 0]
[1 0 0 1 1 1 0 0 1 1]
[4 0 0 0 0 0 0 0 0 0 0]
[2 0 1 0 0 0 1 0 0 0 0 0]
val_acc=0.1953125
--------------------------------------------------------------------------------
[4]
[5 1]
[5 0 0]
[4 0 0 0]
[2 0 0 1 1]
[2 1 0 0 0 0]
[0 0 0 0 0 1 1]
[3 1 0 0 0 0 0 0]
[5 1 1 1 1 0 0 0 1]
[4 0 1 1 1 0 0 1 0 0]
[1 0 0 0 0 0 0 0 1 0 0]
[2 1 1 1 1 1 0 0 0 0 0 0]
val_acc=0.2578125
--------------------------------------------------------------------------------
[0]
[5 0]
[0 0 0]
[4 0 0 1]
[5 0 0 1 1]
[5 0 0 1 0 0]
[2 0 0 0 0 0 0]
[5 1 1 0 0 0 0 1]
[4 1 0 0 0 0 0 0 0]
[0 0 0 1 0 1 0 0 1 0]
[2 1 0 0 1 0 0 0 1 0 0]
[3 1 1 1 0 0 0 0 0 0 0 0]
val_acc=0.28125
--------------------------------------------------------------------------------
[3]
[1 1]
[5 1 0]
[4 0 1 0]
[0 0 0 0 1]
[2 0 0 0 0 0]
[0 0 1 1 0 0 1]
[5 0 0 0 0 0 0 0]
[5 0 0 0 1 0 0 0 0]
[3 0 0 0 0 1 0 1 0 0]
[0 1 0 0 0 1 0 1 0 1 0]
[3 0 0 0 1 1 0 0 0 0 0 0]
val_acc=0.2578125

Epoch 0: Eval
valid_accuracy: 0.2448
test_accuracy: 0.2541
```

**Comment by bellgrylls** [comment]
how the weights sharing is implemented?could you please tell me?I have look around the entire code for servel times

**Comment by TDeVries** [comment]
Each ENAS layer contains the weights for several different branches, and these weights are reused each time we sample a new architecture. When a random path through the architecture is selected it will use the weights for the corresponding branch in each layer. This way even though the overall architecture we sample might be novel, the weights that we use for each branch have been used before.

**Comment by bellgrylls** [comment]
> Each ENAS layer contains the weights for several different branches, and these weights are reused each time we sample a new architecture. When a random path through the architecture is selected it will use the weights for the corresponding branch in each layer. This way even though the overall architecture we sample might be novel, the weights that we use for each branch have been used before.

very helpful,thank you for your excellent implementation!

**Comment by bellgrylls** [comment]
is there something problem with the resume parameter?I resume the program from epoch 50,it show the 'AssertionError: Must define a window to update' error
